### PR TITLE
Improve gRPC CLI

### PR DESCRIPTION
**Pull request by y-zeng**
This PR adds the following improvements to the gRPC command line tool
- Command processing
- `grpc_cli help` command
- CliCrendtials
- Input validation
- Tests for grpc_tool
- Improvements in ProtoFileParser

### CODE REVIEW

File: test/cpp/util/cli_credentials.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+#include <grpc++/security/credentials.h>
+#include <grpc++/support/config.h>
+
+namespace grpc {
+namespace testing {
+
+class CliCredentials {
+ public:
+  virtual std::shared_ptr<grpc::ChannelCredentials> GetCredentials() const;
```
Comment (yang-g) [inline test/cpp/util/cli_credentials.h]: remove virtual or add a virtual dtor?

File: test/cpp/util/proto_file_parser.h [hunk 2]
```diff
@@ -53,41 +51,57 @@ class ProtoFileParser {
   // The given proto file_name will be searched in a source tree rooted from
   // proto_path. The method could be a partial string such as Service.Method or
   // even just Method. It will log an error if there is ambiguity.
```
Comment (yang-g) [inline test/cpp/util/proto_file_parser.h]: Please update comment.

File: test/cpp/util/cli_credentials.h (hunk 1 above)
Comment (y-zeng) [inline test/cpp/util/cli_credentials.h]: Added a virtual dtor. Thanks a lot for the review!

File: test/cpp/util/proto_file_parser.h (hunk 2 above)
Comment (y-zeng) [inline test/cpp/util/proto_file_parser.h]: Fixed.

File: test/cpp/util/cli_credentials.h [hunk 3]
```diff
@@ -42,6 +42,7 @@ namespace testing {
 
 class CliCredentials {
  public:
+  virtual ~CliCredentials() {};
```
Comment (yang-g) [inline test/cpp/util/cli_credentials.h]: redundant semicolon

File: test/cpp/util/proto_file_parser.h [hunk 4]
```diff
@@ -48,15 +48,18 @@ class ErrorPrinter;
...
-  // The given proto file_name will be searched in a source tree rooted from
-  // proto_path. The method could be a partial string such as Service.Method or
-  // even just Method. It will log an error if there is ambiguity.
-  ProtoFileParser(std::shared_ptr<grpc::Channel> channel,
-                  const grpc::string& proto_path,
-                  const grpc::string& protofiles);
+  // The parser will search proto files using the server reflection service
+  // provided on the given channel. The given protofiles in a source tree rooted
+  // from proto_path will also be searched.
+  explicit ProtoFileParser(std::shared_ptr<grpc::Channel> channel,
```
Comment (yang-g) [inline test/cpp/util/proto_file_parser.h]: nit: there is no need of explicit

File: test/cpp/util/cli_credentials.h (hunk 3 above)
Comment (y-zeng) [inline test/cpp/util/cli_credentials.h]: Fixed :sweat_smile:

File: test/cpp/util/proto_file_parser.h (hunk 4 above)
Comment (y-zeng) [inline test/cpp/util/proto_file_parser.h]: Fixed.

**Comment by yang-g** [comment]
LGTM
Thanks for the change.

**Pull request merged**